build(docs-infra): improve directive API doc templates

petebacondarwin opened the pull request:
Closes #22790
Closes #25530

petebacondarwin committed:
build(docs-infra): improve directive API doc templates

Closes #22790
Closes #25530

mary-poppins commented:
You can preview 3db2590 at https://pr25768-3db2590.ngbuilds.io/.

petebacondarwin commented:
Some examples:

https://pr25768-3db2590.ngbuilds.io/api/common/NgClass

<img width="889" alt="screen shot 2018-09-02 at 14 11 18" src="https://user-images.githubusercontent.com/15655/44956330-17212b80-aeba-11e8-8ba4-9627d8d4a610.png">

https://pr25768-3db2590.ngbuilds.io/api/forms/NgForm

![screen shot 2018-09-02 at 14 11 42](https://user-images.githubusercontent.com/15655/44956335-2acc9200-aeba-11e8-825b-79338a9464b3.png)

https://pr25768-3db2590.ngbuilds.io/api/forms/DefaultValueAccessor

![screen shot 2018-09-02 at 14 12 23](https://user-images.githubusercontent.com/15655/44956343-433cac80-aeba-11e8-9c96-7c55e96ca785.png)

https://pr25768-3db2590.ngbuilds.io/api/router/RouterLinkActive

![screen shot 2018-09-02 at 14 13 04](https://user-images.githubusercontent.com/15655/44956347-58194000-aeba-11e8-9a43-d05ff39ec9ee.png)

jenniferfell commented:
notes from mtg 9/6: 
selectors: no table. no indent. UL. 1 LI per selector. 
all tables: no code inside table, very long description or code link to "Usage notes" section
move inputs and outputs into properties table?
idea: separate these?
- properties
- inherited properties
- methods
- inherited methods (inherited from x)
idea: remove class sig from top - everything is described below (for directives)
idea: "selector usage" section instead of generated selector, hand-craft content instead
idea: exports as - "template variable references" is ok. change var to mytemplatevar or myvarname.

petebacondarwin committed:
build(docs-infra): improve directive selector rendering

If the documentation contains a `@selectors` tag then the content of that
is used to describe the selectors of a directive.

Otherwise the selector string is split and each selector is listed as
a list item in an unordered list.

petebacondarwin committed:
build(docs-infra): remove class overview from directive pages

petebacondarwin committed:
build(docs-infra): rename example template variable in directive pages

petebacondarwin committed:
build(docs-infra): directive inputs and outputs

petebacondarwin committed:
build(docs-infra): display inherited members on directives

petebacondarwin committed:
build(docs-infra): include directives etc in class descendants lists

mary-poppins commented:
You can preview d5b4625 at https://pr25768-d5b4625.ngbuilds.io/.

mary-poppins commented:
You can preview 06bdf9b at https://pr25768-06bdf9b.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): move directive macros into memberHelpers.html

petebacondarwin committed:
build(docs-infra): improve directive selector rendering

`:not(...)` blocks are now rendered as italic, while the
rest of the selector is bold.

petebacondarwin committed:
build(docs-infra): pluralize NgModule(s) heading as appropriate

mary-poppins commented:
You can preview 029962b at https://pr25768-029962b.ngbuilds.io/.

mary-poppins commented:
You can preview 0763676 at https://pr25768-0763676.ngbuilds.io/.

mary-poppins commented:
You can preview 42fd7de at https://pr25768-42fd7de.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): add short description "See more" link

If there is additional (non-short) description then add in a
link to the short description to take the reader there.

mary-poppins commented:
You can preview 876f92f at https://pr25768-876f92f.ngbuilds.io/.

gkalpak reviewed (commented):
Random issue:
When a member gets no link (due to disambiguation issues???), then it looks different than the rest ([example][1]):

![animationmetadatatype-rendering-issue](https://user-images.githubusercontent.com/8604205/46015984-4b17f700-c0dc-11e8-88fc-de052ac1780a.png)

[1]: https://pr25768-876f92f.ngbuilds.io/api/animations/AnimationMetadataType#members

IgorMinar commented:
- [ ] in the properties table we need to quote the prop value: `@Input(foo)` -> `@Input("foo")`
- [ ] presentation of getters and setters should be consistent. I think we should actually drop the "get" and "set" prefix for the props and use display only read-only and write-only if necessary. `set foo` -> `foo   write-only`

petebacondarwin committed:
build(docs-infra): remove unused property table heading

petebacondarwin committed:
fixup! build(docs-infra): improve directive API doc templates

petebacondarwin committed:
fixup! build(docs-infra): include directives etc in class descendants lists

petebacondarwin committed:
fixup! build(docs-infra): move directive macros into memberHelpers.html

petebacondarwin committed:
build(docs-infra): simplify property syntax rendering

mary-poppins commented:
You can preview 354b845 at https://pr25768-354b845.ngbuilds.io/.

petebacondarwin commented:
@IgorMinar and @gkalpak PTAL

petebacondarwin committed:
build(docs-infra): fix formatting of entry point export table

Now that `list-table` cells are `pre` formatterd we must be careful
of what whitespace appears in text nodes.

petebacondarwin commented on aio/src/styles/2-modules/_api-pages.scss:
```diff
@@ -33,6 +33,7 @@
   .method-table, .option-table, .list-table {
     td > code {
       background-color: inherit;
+      white-space: pre;
```
fixed in 8d644eec178199d4629d9c4fd0e55b81586635ba

mary-poppins commented:
You can preview 8d644ee at https://pr25768-8d644ee.ngbuilds.io/.

gkalpak commented on aio/tools/transforms/angular-base-package/rendering/hasValues.js:
```diff
@@ -8,9 +8,16 @@ module.exports = function hasValues() {
 function readProperty(obj, propertySegments, index) {
   const value = obj[propertySegments[index]];
-  return index < (propertySegments.length - 1) ? readProperty(value, propertySegments, index + 1) : value;
+  return !!value && (index === propertySegments.length - 1 || readProperty(value, propertySegments, index + 1));
```
Not sure if it is intentional, but I just realized this will return false if the target property exists but is falsy. E.g. `readProperty(obj, ['a', 'b'], 0)` will return true for `obj = {a: {b: 1}}` but false for `obj = {a: {b: 0}}`.
(This is possibly the intended behavior, though.)

petebacondarwin replied:
Yes this is intended because we don't want `""` to count as a real value - and so only truthy values are significant here.

gkalpak commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -191,8 +189,9 @@ <h2>{$ headingText $}</h2>
-          {% if property.shortDescription %}{$ property.shortDescription | marked $}{% endif %}
+            {%- if (property.isGetAccessor or property.isReadonly) and not property.isSetAccessor %}<span class='read-only-property'>Read-only.</span>{% endif %}
+            {%- if property.isSetAccessor and not property.isGetAccessor %}<span class='write-only-property'>Write-only.</span>{% endif %}
+            {% if property.shortDescription %}{$ property.shortDescription | marked $}{% endif %}
```
Incorrect indentation 😱

petebacondarwin committed:
fixup! build(docs-infra): simplify property syntax rendering

petebacondarwin committed:
fixup! build(docs-infra): improve directive API doc templates

mary-poppins commented:
You can preview b09ba76 at https://pr25768-b09ba76.ngbuilds.io/.

mary-poppins commented:
You can preview a5d9691 at https://pr25768-a5d9691.ngbuilds.io/.

petebacondarwin committed:
fixup! build(docs-infra): simplify property syntax rendering

mary-poppins commented:
You can preview d882c07 at https://pr25768-d882c07.ngbuilds.io/.

petebacondarwin committed:
fix(docs-infra): remove unnecessary margin on short descriptions

(This was added in 405d97431fa374afdd5cddc6780f8949ee76a2c6 but it is
not clear the reasoning. It looks better to remove it now.)

mary-poppins commented:
You can preview 9772127 at https://pr25768-9772127.ngbuilds.io/.

IgorMinar reviewed (approved):
I think we still need to tweak this further, but let's get this in a build on top of it.
